appvalidate/coastal_secrets.go

```diff
@@ -0,0 +1,210 @@
+	if !ok {
+		return nil, fmt.Errorf("unexpected coastal private key type: %T", privKeyAny)
+	}
+	return &CoastalSecrets{
+		Config: *c,
+		pk:     privateKey,
+	}, nil
+}
+
+// CoastalPayload is the external front facing payload sent to platform from clients.
+type CoastalPayload struct {
+	RCTypeID        string `json:"rcTypeId"`
+	RCInstanceID    string `json:"rcInstanceId"`
+	HardwareVersion string `json:"rcHWVersion"`
+	SoftwareVersion string `json:"rcSWVersion"`
+	PHDTypeID       string `json:"phdTypeId"`
+	PHDInstanceID   string `json:"phdInstanceId"`
+	CSR             string `json:"csr"`
+}
+
+func (c *CoastalPayload) toInternalPayload(pk ed25519.PrivateKey) (payload *coastalPayload, signature string, err error) {
+	payload = &coastalPayload{
+		RCTypeID:         c.RCTypeID,
+		RCInstanceID:     c.RCInstanceID,
+		HardwareVersions: []string{c.HardwareVersion},
+		SoftwareVersions: []string{c.SoftwareVersion},
+		PHDTypeID:        c.PHDTypeID,
+		PHDInstanceID:    c.PHDInstanceID,
+		CSR:              c.CSR,
+	}
+
+	bytesRaw, err := json.Marshal(payload)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to marshal payload when calculating signature: %w", err)
+	}
+	signature = base64.StdEncoding.EncodeToString(ed25519.Sign(pk, bytesRaw))
+	return payload, signature, nil
+}
+
+// coastalPayload is the internal facing actual Coastal payload that platform sends to coastal's API.
+type coastalPayload struct {
+	RCTypeID         string   `json:"rcTypeId"`
+	RCInstanceID     string   `json:"rcInstanceId"`
+	HardwareVersions []string `json:"rcHWVersions"`
+	SoftwareVersions []string `json:"rcSWVersions"`
+	PHDTypeID        string   `json:"phdTypeId"`
+	PHDInstanceID    string   `json:"phdInstanceId"`
+	CSR              string   `json:"csr"`
+}
+
+type CoastalResponse struct {
+	Certificates []struct {
+		Content   string `json:"content"`
+		TTLInDays int    `json:"ttlInDays"`
+		Type      string `json:"type"`
+	} `json:"certificates"`
+}
+
+func (c *CoastalSecrets) GetSecret(ctx context.Context, partnerDataRaw []byte) (*CoastalResponse, error) {
+	if c.pk == nil {
+		return nil, ErrInvalidPartnerCredentials
+	}
+	payload := newCoastalPayload(c.Config.RCTypeID)
+	if err := json.Unmarshal(partnerDataRaw, payload); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal Coastal payload: %w", err)
+	}
+
+	if err := structValidator.New().Validate(payload); err != nil {
+		return nil, fmt.Errorf("Coastal: %w: %w", ErrInvalidPartnerPayload, err)
+	}
+
+	internalPayload, signature, err := payload.toInternalPayload(c.pk)
+	if err != nil {
+		return nil, fmt.Errorf("unable to genarate internal Coastal payload: %w", err)
+	}
+
+	b, err := json.Marshal(internalPayload)
+	if err != nil {
+		return nil, fmt.Errorf("Coastal marshal error: %w", err)
+	}
+
+	u, err := url.Parse(c.Config.BaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse Coastal API baseURL: %w", err)
+	}
+	u.Path = path.Join(u.Path, "devices/api/v1/certificates")
```

Comment:
This could be done once in `newCoastalSecretsConfig` instead, and just stored as a string there.